### PR TITLE
optimize: use klauspost/compress

### LINKED ISSUE #343: Use better gzip/deflate
The standard library is more interested in maintaining status quo than improving compression.

I maintain a [compression package](https://github.com/klauspost/compres) that seems more suited to Hertz. It plugs in without any modifications, by simply replacing imports.

Below is a comparison between my package and the standard library, using [4GB of mixed web content](https://files.klauspost.com/compress/webdevdata.org-2015-01-07-4GB-subset.7z):

|  out  | level | files |     insize |    outsize | millis |   mb/s | Files/s | MB/s   | Reduction | Rel speed | Smaller | Gain    |
|:-----:|------:|------:|-----------:|-----------:|-------:|-------:|---------|--------|-----------|-----------|---------|---------|
| gzstd |     1 | 53927 | 4014735833 |  862232427 |  20605 | 185.81 |    2617 | 185.82 |    78.52% |     1.00x |  0.000% |  0.000% |
| gzkp  |     1 | 53927 | 4014735833 |  866657631 |  13860 | 276.24 |    3891 | 276.24 |    78.41% |     1.49x | -0.513% | -0.110% |
|       |       |       |            |            |        |        |         |        |           |           |         |         |
| gzstd |     2 | 53927 | 4014735833 |  813642417 |  26380 | 145.14 |    2044 | 145.14 |    79.73% |     1.00x |  0.000% |  0.000% |
| gzkp  |     2 | 53927 | 4014735833 |  842584206 |  15322 | 249.87 |    3520 | 249.89 |    79.01% |     1.72x | -3.557% | -0.721% |
|       |       |       |            |            |        |        |         |        |           |           |         |         |
| gzstd |     3 | 53927 | 4014735833 |  800582015 |  27842 | 137.51 |    1937 | 137.52 |    80.06% |     1.00x |  0.000% |  0.000% |
| gzkp  |     3 | 53927 | 4014735833 |  823209525 |  17003 | 225.17 |    3172 | 225.18 |    79.50% |     1.64x | -2.826% | -0.564% |
|       |       |       |            |            |        |        |         |        |           |           |         |         |
| gzstd |     4 | 53927 | 4014735833 |  765294480 |  34710 | 110.31 |    1554 | 110.31 |    80.94% |     1.00x |  0.000% |  0.000% |
| gzkp  |     4 | 53927 | 4014735833 |  791607893 |  19778 | 193.58 |    2727 | 193.59 |    80.28% |     1.75x | -3.438% | -0.655% |
|       |       |       |            |            |        |        |         |        |           |           |         |         |
| gzstd |     5 | 53927 | 4014735833 |  744808940 |  42672 |  89.72 |    1264 |  89.73 |    81.45% |     1.00x |  0.000% |  0.000% |
| gzkp  |     5 | 53927 | 4014735833 |  764683084 |  22471 | 170.38 |    2400 | 170.39 |    80.95% |     1.90x | -2.668% | -0.495% |
|       |       |       |            |            |        |        |         |        |           |           |         |         |
| gzstd |     6 | 53927 | 4014735833 |  735318526 |  51993 |  73.64 |    1037 |  73.64 |    81.68% |     1.00x |  0.000% |  0.000% |
| gzkp  |     6 | 53927 | 4014735833 |  753831382 |  24787 | 154.46 |    2176 | 154.47 |    81.22% |     2.10x | -2.518% | -0.461% |
|       |       |       |            |            |        |        |         |        |           |           |         |         |
| gzstd |     7 | 53927 | 4014735833 |  733144257 |  60588 |  63.19 |     890 |  63.19 |    81.74% |     1.00x |  0.000% |  0.000% |
| gzkp  |     7 | 53927 | 4014735833 |  746309058 |  34562 | 110.78 |    1560 | 110.78 |    81.41% |     1.75x | -1.796% | -0.328% |
|       |       |       |            |            |        |        |         |        |           |           |         |         |
| gzstd |     8 | 53927 | 4014735833 |  730369312 |  77883 |  49.16 |     692 |  49.16 |    81.81% |     1.00x |  0.000% |  0.000% |
| gzkp  |     8 | 53927 | 4014735833 |  735523067 |  43270 |  88.48 |    1246 |  88.49 |    81.68% |     1.80x | -0.706% | -0.128% |
|       |       |       |            |            |        |        |         |        |           |           |         |         |
| gzstd |     9 | 53927 | 4014735833 |  729905331 |  94517 |  40.51 |     571 |  40.51 |    81.82% |     1.00x |  0.000% |  0.000% |
| gzkp  |     9 | 53927 | 4014735833 |  727900632 |  94305 |   40.6 |     572 |  40.60 |    81.87% |     1.00x |  0.275% |  0.050% |
|       |       |       |            |            |        |        |         |        |           |           |         |         |
| gzstd |    -2 | 53927 | 4014735833 | 2617482950 |  12742 | 300.48 |    4232 | 300.48 |    34.80% |     1.00x |  0.000% |  0.000% |
| gzkp  |    -2 | 53927 | 4014735833 | 2604521759 |   9281 | 412.49 |    5810 | 412.54 |    35.13% |     1.37x |  0.495% |  0.323% |

The standard library is poorly balanced, so levels 6-8 have very little difference. This is more spread out so the actual level reflects an increase in compression. Here is the compression/speed tradeoff visualized:

chart

The package uses no cgo or "unsafe" code.

**Describe the solution you'd like**

Replace gzip/flate imports with [corresponding packages](https://github.com/klauspost/compress#deflate-usage).

**Additional context**

The package also provides [stateless compression modes](https://github.com/klauspost/compress#stateless-compression) which can be useful for long-running requests with little data throughput.

Each Write will be compressed immediately without any state being retained between calls. I don't know if this has any use here.

### PR DESCRIPTION
**Pull request by li-jin-gou**
#### What type of PR is this?

perf

#### What this PR does / why we need it (English/Chinese):
- zh: 使用 klauspost/compress

#### Which issue(s) this PR fixes:

Fixes https://github.com/cloudwego/hertz/issues/343

### CODE REVIEW

**Review by klauspost — APPROVED** [review]
LGTM

File: example/main.go [hunk 1]
```diff
@@ -1,27 +1,42 @@
...
-// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
-// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
-// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
-// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
-// SOFTWARE.
-//
-// This file may have been modified by CloudWeGo authors. All CloudWeGo
-// Modifications are Copyright 2022 CloudWeGo Authors.
+/*
+ * Copyright 2022 CloudWeGo Authors
```
Comment (welkeyever) [inline example/main.go]: 2023？

File: gzip.go [hunk 2]
```diff
@@ -1,34 +1,48 @@
...
-// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
-// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
-// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
-// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
-// SOFTWARE.
-//
-// This file may have been modified by CloudWeGo authors. All CloudWeGo
-// Modifications are Copyright 2022 CloudWeGo Authors.
+/*
+ * Copyright 2022 CloudWeGo Authors
```
Comment (welkeyever) [inline gzip.go]: 2023, same with others

**Review by L2ncE — APPROVED** [review]

**Review by justlorain — APPROVED** [review]

File: options.go [hunk 3]
```diff
@@ -84,7 +99,7 @@ func WithDecompressFn(decompressFn app.HandlerFunc) Option {
 }
 
 func NewExcludedPaths(paths []string) ExcludedPaths {
-	return ExcludedPaths(paths)
+	return paths
```
Comment (welkeyever) [inline options.go:102]: Why change this?

**Comment by welkeyever** [comment]
Adding this as an option is better than replacing the whole impl.

File: options.go (hunk 3 above)
Comment (li-jin-gou) [reply options.go:102]:
```
type (
	Options struct {
		ExcludedExtensions  ExcludedExtensions
		ExcludedPaths       ExcludedPaths
		ExcludedPathRegexes ExcludedPathRegexes
		DecompressFn        app.HandlerFunc
	}
	Option              func(*Options)
	ExcludedExtensions  map[string]bool
	ExcludedPaths       []string
	ExcludedPathRegexes []*regexp.Regexp
)
```

ExcludedPaths  is []string

**Pull request closed without merging**